Improved MyBatis support.

schauder opened the pull request:
Delegating auto configuration of MyBatis to the official MyBatis Boot Starter.

Corrected (the unfortunate complex) construction of DataAccessStrategies.

Configured the Maven Surefire Plugin to actually execute the tests.

schauder committed:
Improved MyBatis support.

Delegating auto configuration of MyBatis to the official MyBatis Boot Starter.

Corrected (the unfortunate complex) construction of DataAccessStrategies.

Configured the Maven Surefire Plugin to actually execute the tests.

gregturn commented on src/main/java/org/springframework/boot/autoconfigure/data/jdbc/JdbcRepositoriesMyBatisAutoConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+	@Bean("dataAccessStrategy")
+	@ConditionalOnMissingBean
+	DataAccessStrategy dataAccessStrategyWithMyBatisSupport(NamedParameterJdbcOperations operations,
+			SqlGeneratorSource sqlGeneratorSource, JdbcMappingContext context, SqlSessionFactory sqlSessionFactory) {
```
Not sure why you need to name the bean anything in particular. I _think_ if you name the method `dataAccessStrategy`, it will give the bean that name.

schauder replied:
I'm pretty sure the assumption is correct. Was from the version where there where two methods returning a DataAccessStrategy in the same class.

schauder replied:
I want the name to be "dataAccessStrategy" just to be consistent with common practice, just in case someone wants to access by name.

gregturn commented on src/main/java/org/springframework/boot/autoconfigure/data/jdbc/JdbcRepositoriesMyBatisAutoConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+ */
+@Configuration
+@ConditionalOnClass(SqlSessionFactory.class)
+@Import(MybatisAutoConfiguration.class)
```
This should be unnecessary. Any external autoconfiguration policies are responsible for wiring themselves. Perhaps you actually mean `@AutoConfigureAfter(MybaticAutoConfiguration.class)`, which would be proper.

To make this work in unit tests, add `MybatisAutoConfiguration` to the `prepare()` setup.

schauder replied:
That's something a thought about and couldn't really make up my mind: Should one depend on the starter and let autoconfig make its thing? Or just depend on the autoconfig and call/import it. ... so I guess you are voting for the first.

schauder replied:
I'm not including the MybatisAutoConfiguration in the tests, since that would require a full configuration of MyBatis (specifying Mappers and stuff).

gregturn commented on src/main/java/org/springframework/boot/autoconfigure/data/jdbc/JdbcRepositoriesMyBatisAutoConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+		System.out.println("configuring mybatis");
+		return JdbcRepositoriesAutoConfiguration.buildDataAccessStrategy(new MyBatisDataAccessStrategy(sqlSessionFactory),
+				new DefaultDataAccessStrategy(sqlGeneratorSource, operations, context));
+	}
```
If we're using MyBatis, why do we need the other one? Is this in the event you only use MyBatis for perhaps one particular aggregate?

schauder replied:
With MyBatis statements may come from the `SqlSessionFactory`. If it doesn't contain the query in question we still fall back on the `DefaultDataAccessStrategy`

gregturn commented on src/test/java/org/springframework/boot/autoconfigure/data/jdbc/JdbcRepositoriesMyBatisAutoConfigurationTests.java:
```diff
@@ -47,11 +51,29 @@ public void tearDown() {
+		DataAccessStrategy dataAccessStrategy = this.context.getBean(DataAccessStrategy.class);
+		List strategies = getField(dataAccessStrategy, "strategies");
+		assertThat(strategies).hasSize(2);
 
```
I would favor peeking at the collection and asserting that `MyBatisDataAccessStrategy` is in there.

gregturn commented on src/test/java/org/springframework/boot/autoconfigure/data/jdbc/support/TestUtilities.java:
```diff
@@ -41,4 +46,16 @@ public static AnnotationConfigApplicationContext prepareApplicationContext(Class
+		try {
+
+			Field field = CascadingDataAccessStrategy.class.getDeclaredField(name);
+			field.setAccessible(true);
```
Seems like a handy utility, however, it's immediately confined to a single class type. Consider making the class a third argument?

schauder replied:
Oh ... missed that.

gregturn reviewed (commented):
Looks good. One thing, I wouldn't get too wrapped up in formatting Spring Data style, given this code will ultimately be a part of Boot and get reformatted in their style. That's why I haven't commented on any sort of polishing.

Additionally, I'm in favor of switching on travisCI and enabling publishing to artifactory, making it available for consumption. I'm willing to execute that come Monday.

schauder committed:
Incorporating feedback from review.

Not importing the mybatis autoconfiguration, but leaving that the starter which is now a dependency.

Made TestUtilities.getField work for arbitrary source classes.

Made tests for construction of DataAccessStrategies more explicit.

schauder committed:
Getting travis to build the project

gregturn commented:
Merge at will!
